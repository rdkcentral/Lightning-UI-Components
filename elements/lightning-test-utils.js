/**
 * Copyright 2020 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
import { readFileSync } from 'fs';
import TestRenderer from './lightning-test-renderer';

export default {
  fastForward,
  makeCreateComponent,
  pathToDataURI
};

function fastForward(elements) {
  const _fastForward = element => {
    if (element._transitions) {
      Object.values(element._transitions).forEach(transition =>
        transition.finish()
      );
    }
  };
  if (typeof elements.forEach === 'function') elements.forEach(_fastForward);
  if (elements && elements._transitions) _fastForward(elements);
}

function pathToDataURI(path) {
  return `data:image/jpg;base64, ${readFileSync(path).toString('base64')}`;
}

function makeCreateComponent(type, defaultConfig = {}, options = {}) {
  return (config = {}) => {
    const testRenderer = TestRenderer.create(
      {
        Component: {
          type,
          ...defaultConfig,
          ...config
        }
      },
      options
    );
    return [testRenderer.getInstance(), testRenderer];
  };
}
