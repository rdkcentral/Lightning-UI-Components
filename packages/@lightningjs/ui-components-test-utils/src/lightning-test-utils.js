/**
 * Copyright 2023 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { readFileSync } from 'fs';
import sinon from 'sinon';
import { context } from '@lightningjs/ui-components';
import TestRenderer from './lightning-test-renderer.js';

export default {
  fastForward,
  makeCreateComponent,
  pathToDataURI,
  nextTick,
  completeAnimation
};

export { TestRenderer };

export function nextTick(wait = 0) {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve();
    }, wait);
  });
}

/**
 * Extends a component and provides methods that use Jest spies to create
 * mock functions for each method passed to the methods param. Each spy method
 * created is a Promise, which resolves after the spied upon method is invoked,
 * and can be awaited in tests.
 * @param { Component } Component component with methods to be mocked
 * @param { string[] } methods names of methods to be mocked on Component
 *
 * @return {lng.Component} a child class of Component with added mock function
 * methods that can be awaited in test cases.
 */
function withMethodSpies(Component, methods, jest) {
  return class SpyClass extends Component {
    static get name() {
      return Component.name;
    }

    _construct() {
      super._construct && super._construct();

      // Create a promise to be resolved following a method being invoked
      const generateSpyPromise = spyResolver => {
        return new Promise(resolve => {
          this[spyResolver] = resolve;
        });
      };

      // Reassign promise in mock function to a new pending Promise
      const resetSpyPromise = (spyPromise, spyResolver) => {
        this[spyPromise] = generateSpyPromise(spyResolver);
      };

      methods.forEach(method => {
        const spyMethod = SpyClass.prototype[method];
        const spyPromise = `_${method}SpyPromise`;
        const spyResolver = `_${method}SpyResolve`;

        // Add Promise which resolves following the mock function being called
        this[spyPromise] = generateSpyPromise(spyResolver);

        // Create mock function which calls the class method and resolves a
        // Promise that can be awaited in test cases

        if (jest) {
          jest
            .spyOn(SpyClass.prototype, method)
            .mockImplementation(async function (...args) {
              await spyMethod.call(this, ...args);
              this[spyResolver]();
              resetSpyPromise(spyPromise, spyResolver);
            });
        } else {
          sinon
            .stub(SpyClass.prototype, method)
            .callsFake(async function (...args) {
              await spyMethod.call(this, ...args);
              this[spyResolver]();
              resetSpyPromise(spyPromise, spyResolver);
            });
        }
      });
    }
  };
}

export function fastForward(elements) {
  const _fastForward = element => {
    if (element._transitions) {
      Object.values(element._transitions).forEach(transition =>
        transition.finish()
      );
    }
  };
  if (typeof elements.forEach === 'function') elements.forEach(_fastForward);
  if (elements && elements._transitions) _fastForward(elements);
}

/**
 * // TODO: this doesn't currently work with monorepo.
 * The path passed in won't lead to anything unless that same asset
 * is in both @lightningjs/tests-utils as well as the package from where
 * the test is called from.
 */
export function pathToDataURI(path) {
  return `data:image/jpg;base64, ${readFileSync(path).toString('base64')}`;
}

export function makeCreateComponent(
  type,
  defaultConfig = {},
  defaultOptions = {},
  defaultGlobals = {}
) {
  return (config = {}, options = {}, instanceGlobals = {}) => {
    const globals = { ...defaultGlobals, ...instanceGlobals };
    const testRenderer = TestRenderer.create(
      {
        Component: {
          type: options.spyOnMethods
            ? withMethodSpies(type, options.spyOnMethods, globals.jest)
            : type,
          ...defaultConfig,
          ...config
        }
      },
      { ...defaultOptions, ...options }
    );
    return [testRenderer.getInstance(), testRenderer];
  };
}

/**
 * Returns a Promise that resolves once all animating properties have updated
 * to their target value(s).
 * @param { Element } Element with properties that have animations
 * @param { string | string[] } transitionProperties property name or names
 * that will be animating.
 *
 * @return {Promise} Promise which resolves upon completion of animations of
 * transitionProperties have completed.
 */
export function completeAnimation(element, transitionProperties = []) {
  const props = Array.isArray(transitionProperties)
    ? transitionProperties
    : [transitionProperties];
  const transitions = props.map(prop => {
    return new Promise(resolve => {
      element._getTransition(prop).once('finish', () => {
        resolve();
      });
    });
  });

  return Promise.all(transitions);
}

/**
 * This function will set up context for the test suite
 * Should be used when testing the context is required.
 * @param mockKeyMetricsHandler a user provided mock handler
 * @return {{keyMetricsCallback: jest.Mock}}
 */
// const mockContext = (context, mockKeyMetricsHandler = jest.fn()) => {
const mockContext = (context, mockKeyMetricsHandler = sinon.spy()) => {
  context.config({
    keyMetricsCallback: mockKeyMetricsHandler
  });

  return context;
};

const resetContext = () => {
  context.setKeyMetricsCallback = undefined;
};

export { mockContext, resetContext };
