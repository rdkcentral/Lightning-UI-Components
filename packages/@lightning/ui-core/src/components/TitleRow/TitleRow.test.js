import lng from '@lightningjs/core';
import { makeCreateComponent } from '../../../test/lightning-test-utils.js';
import { TitleRow } from '../';

const items = Array.from({ length: 7 }).map(() => ({
  type: lng.Component,
  w: 300,
  h: 200
}));

const createComponent = makeCreateComponent(TitleRow);

describe('TitleRow', () => {
  let component, testRenderer;

  beforeEach(() => {
    [component, testRenderer] = createComponent({
      items
    });
  });

  it('renders', () => {
    [component, testRenderer] = createComponent({
      title: 'Title',
      items
    });
    testRenderer.forceAllUpdates();
    const tree = testRenderer.toJSON(2);

    expect(tree).toMatchSnapshot();
  });

  it('sets the announce string to the row title text', () => {
    const title = 'Row Title';
    component.title = title;
    testRenderer.forceAllUpdates();
    expect(component.announce).toBe(title);
  });

  it('overrides the announce string', () => {
    const overrideString = 'Custom announce string';
    component.announce = overrideString;
    testRenderer.forceAllUpdates();
    expect(component.announce).toBe(overrideString);
  });

  describe('updating the title', () => {
    it('should add and update a title', () => {
      const titleA = 'Title';
      const titleB = 'New Title';

      expect(component._Title.content).toBe('');

      component.title = titleA;

      testRenderer.forceAllUpdates();

      expect(component._Title.content).toBe(titleA);

      component.title = titleB;
      testRenderer.forceAllUpdates();

      expect(component._Title.content).toBe(titleB);
    });
    it('should remove a title', () => {
      const title = 'Title';
      component.title = title;
      testRenderer.forceAllUpdates();
      expect(component._Title.content).toBe(title);
      component.title = undefined;
      testRenderer.forceAllUpdates();
      expect(component._Title).toMatchObject({
        h: 0,
        content: ''
      });
    });
    it('should update the row placement after the title has loaded', async () => {
      [component, testRenderer] = createComponent(
        {
          title: 'Title',
          items
        },
        {
          spyOnMethods: ['_titleLoaded']
        }
      );
      expect(component._Items.y).toBe(component.style.rowMarginTop);

      await component.__titleLoadedSpyPromise;

      expect(component._Items.y).toBe(
        component._Title.finalH + component.style.rowMarginTop
      );
    });
  });
});
