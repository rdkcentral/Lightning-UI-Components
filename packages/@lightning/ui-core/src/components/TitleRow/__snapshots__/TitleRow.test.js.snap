// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TitleRow renders 1`] = `
{
  "Component": {
    "Items": {
      "active": true,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": {
        "Element-16": {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 200,
          "hasFinalFocus": true,
          "hasFocus": true,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": null,
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "type": "Component",
          "visible": true,
          "w": 300,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
        "Element-17": {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 200,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": null,
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "type": "Component",
          "visible": true,
          "w": 300,
          "x": 324,
          "y": 0,
          "zIndex": 0,
        },
        "Element-18": {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 200,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": null,
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "type": "Component",
          "visible": true,
          "w": 300,
          "x": 648,
          "y": 0,
          "zIndex": 0,
        },
        "Element-19": {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 200,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": null,
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "type": "Component",
          "visible": true,
          "w": 300,
          "x": 972,
          "y": 0,
          "zIndex": 0,
        },
        "Element-20": {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 200,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": null,
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "type": "Component",
          "visible": true,
          "w": 300,
          "x": 1296,
          "y": 0,
          "zIndex": 0,
        },
        "Element-21": {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 200,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": null,
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "type": "Component",
          "visible": true,
          "w": 300,
          "x": 1620,
          "y": 0,
          "zIndex": 0,
        },
        "Element-22": {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 200,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": null,
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "type": "Component",
          "visible": true,
          "w": 300,
          "x": 1944,
          "y": 0,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 200,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Items",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": [
        "Items",
      ],
      "visible": true,
      "w": 2244,
      "x": 0,
      "y": 72,
      "zIndex": 0,
    },
    "Title": {
      "active": true,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": {
        "Text": {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 0,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Text",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": [
            "Text",
          ],
          "texture": {
            "fontFace": "XfinityBrownBold",
            "fontSize": 36,
            "letterSpacing": -0.2,
            "lineHeight": 48,
            "maxLines": undefined,
            "maxLinesSuffix": "...",
            "text": "Title",
            "textBaseline": "bottom",
            "textColor": 4294375161,
            "textOverflow": "ellipsis",
            "type": "TextTexture",
            "verticalAlign": "middle",
            "wordWrapWidth": undefined,
          },
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 2,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 48,
      "hasFinalFocus": false,
      "hasFocus": false,
      "isComponent": true,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Title",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": "",
      "tag": [Function],
      "tags": [
        "Title",
      ],
      "type": "TextBox",
      "visible": true,
      "w": 4,
      "x": 24,
      "y": 0,
      "zIndex": 0,
    },
  },
}
`;
